src/backend/web-gl/kernel.js

```diff
@@ -700,8 +700,21 @@ module.exports = class WebGLKernel extends KernelBase {
 						length *= 4;
 					}
 
-					const valuesFlat = new Float32Array(length);
-					utils.flattenTo(value, valuesFlat);
+					let valuesFlat
+
+					if (utils.isArray(value[0])) {
+						// not already flat
+						valuesFlat = new Float32Array(length);
+						utils.flattenTo(value, valuesFlat);
+					} else if (value.constructor != Float32Array) {
+						// TODO: would be great if we could not have to create Float32Array buffers
```

Comment:
Should we open an issue?

Reply:
could do, was what we were talking about yesterday!

Reply:
https://github.com/gpujs/gpu.js/issues/346

ty @nickhingston!!!!